chainimport/headers_import.go

```diff
@@ -0,0 +1,1403 @@
+	result.Duration = result.EndTime.Sub(result.StartTime)
+
+	log.Infof("Headers import completed: processed %d headers "+
+		"(block and filter) (added: %d, skipped: %d) from height "+
+		"%d to %d in %s (%.2f headers/sec, %.2f%% new)",
+		result.ProcessedCount, result.AddedCount, result.SkippedCount,
+		result.StartHeight, result.EndHeight,
+		formatDuration(result.Duration), result.HeadersPerSecond(),
+		result.NewHeadersPercentage())
+
+	return result, nil
+}
+
+// isTargetFresh checks if the target header stores are in their initial state,
+// meaning they contain only the genesis header (height 0).
+func (s *HeadersImport) isTargetFresh(
+	targetBlockHeaderStore headerfs.BlockHeaderStore,
+	targetFilterHeaderStore headerfs.FilterHeaderStore) (bool, error) {
+
+	// Get the chain tip from both target stores.
+	_, blockTipHeight, err := targetBlockHeaderStore.ChainTip()
+	if err != nil {
+		return false, fmt.Errorf("failed to get target block header "+
+			"chain tip: %w", err)
+	}
+
+	_, filterTipHeight, err := targetFilterHeaderStore.ChainTip()
+	if err != nil {
+		return false, fmt.Errorf("failed to get target filter header "+
+			"chain tip: %w", err)
+	}
+
+	if blockTipHeight == 0 && filterTipHeight == 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (s *HeadersImport) initSources() error {
+	// Check if required import sources are provided.
+	if s.BlockHeadersImportSource == nil || s.FilterHeadersImportSource == nil {
+		return fmt.Errorf("missing required header sources - block "+
+			"headers source: %v, filter headers source: %v",
+			s.BlockHeadersImportSource != nil,
+			s.FilterHeadersImportSource != nil)
+	}
+
+	// Check if required validators are provided.
+	if s.BlockHeadersValidator == nil || s.FilterHeadersValidator == nil {
+		return fmt.Errorf("missing required header validators - block "+
+			"headers validator: %v, filter headers "+
+			"validator: %v", s.BlockHeadersValidator != nil,
+			s.FilterHeadersValidator != nil)
+	}
+
+	// Init Block Headers Import Source.
+	if err := s.BlockHeadersImportSource.Open(); err != nil {
+		return err
+	}
+
+	// Init Filter Headers Import Source.
+	if err := s.FilterHeadersImportSource.Open(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// deinitSources safely closes all open header sources and logs any warnings
+// encountered during cleanup.
+func (s *HeadersImport) deinitSources() {
+	// Close block headers source.
+	if err := s.BlockHeadersImportSource.Close(); err != nil {
+		log.Warnf("Failed to close block headers source: %v", err)
+	}
+
+	// Close filter headers source.
+	if err := s.FilterHeadersImportSource.Close(); err != nil {
+		log.Warnf("Failed to close block headers source: %v", err)
+	}
+
+	// Uinitialize logger.
+	log = nil
+}
+
+// heightToSrcIndex converts an absolute blockchain height to a
+// source-relative index based on the provided start height.
+func (s *HeadersImport) heightToSrcIndex(height, startHeight uint32) int {
```

Comment:
This can be a pure function, no need to be a method. Then it can be moved to the bottom of the file, so it's easier to follow all the actual methods of `HeadersImport`.